Firestore/core/src/firebase/firestore/remote/grpc_unary_call.cc

```diff
@@ -0,0 +1,106 @@
+/*
+ * Copyright 2018 Google
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "Firestore/core/src/firebase/firestore/remote/grpc_unary_call.h"
+
+#include "Firestore/core/src/firebase/firestore/remote/grpc_connection.h"
+
+#include <utility>
+
+namespace firebase {
+namespace firestore {
+namespace remote {
+
+using util::AsyncQueue;
+using util::Status;
+using Tag = GrpcCompletion::Tag;
+
+GrpcUnaryCall::GrpcUnaryCall(
+    std::unique_ptr<grpc::ClientContext> context,
+    std::unique_ptr<grpc::GenericClientAsyncResponseReader> call,
+    AsyncQueue* worker_queue,
+    GrpcConnection* grpc_connection,
+    const grpc::ByteBuffer& request)
+    : context_{std::move(context)},
+      call_{std::move(call)},
+      request_{request},
+      worker_queue_{worker_queue},
+      grpc_connection_{grpc_connection} {
+  grpc_connection_->Register(this);
+}
+
+GrpcUnaryCall::~GrpcUnaryCall() {
+  HARD_ASSERT(!finish_completion_,
+              "GrpcUnaryCall is being destroyed without proper shutdown");
+}
+
+void GrpcUnaryCall::Start(CallbackT&& callback) {
+  callback_ = std::move(callback);
+  call_->StartCall();
+
+  finish_completion_ = new GrpcCompletion(
```

Comment:
Nothing to do with your code, but this has always struck me as a slightly weird thing with grpc++. We're allocating memory here, and then relying on the completion handler to cast it back to the original type, do something with it, and free the memory. It's hard to track this memory via inspection unless you know the pattern that's being used.

This *might* be worth a comment indicating how the memory is handled?

Reply:
Added a "pointer" comment (`grpc_completion.h` describes the completion's lifetime, added a comment referring to that file's documentation).